tests/pkg/client/client.go

```diff
@@ -11,31 +16,52 @@ import (
 	"k8s.io/client-go/tools/clientcmd"
 )
 
+var (
+	masterUrl      string
+	kubeconfigPath string
+)
+
+func init() {
+	flag.StringVar(&kubeconfigPath, "kubeconfig", "",
+		"path to a kubeconfig. Only required if out-of-cluster.")
+	flag.StringVar(&masterUrl, "master", "",
+		"address of the Kubernetes API server. Overrides any value in kubeconfig. "+
+			"Only required if out-of-cluster.")
+}
+
 func NewCliOrDie() (versioned.Interface, kubernetes.Interface) {
-	cfg, err := rest.InClusterConfig()
+	cfg, err := GetConfig()
 	if err != nil {
 		panic(err)
 	}
 
-	cfg.Timeout = 30 * time.Second
-	cli, err := versioned.NewForConfig(cfg)
-	if err != nil {
-		panic(err)
-	}
+	return buildClientsOrDie(cfg)
+}
 
-	kubeCli, err := kubernetes.NewForConfig(cfg)
-	if err != nil {
-		panic(err)
+func GetConfig() (*rest.Config, error) {
+	// If kubeconfigPath provided, use that
+	if len(kubeconfigPath) > 0 {
+		return clientcmd.BuildConfigFromFlags(masterUrl, kubeconfigPath)
+	}
+	// If an env variable is specified with the config locaiton, use that
+	if len(os.Getenv("KUBECONFIG")) > 0 {
+		return clientcmd.BuildConfigFromFlags(masterUrl, os.Getenv("KUBECONFIG"))
+	}
+	// If no explicit location, try the in-cluster config
+	if c, err := rest.InClusterConfig(); err == nil {
+		return c, nil
+	}
+	// If no in-cluster config, try the default location in the user's home directory
+	if usr, err := user.Current(); err == nil {
+		if c, err := clientcmd.BuildConfigFromFlags(
+			"", filepath.Join(usr.HomeDir, ".kube", "config")); err == nil {
+			return c, nil
```

Comment:
This operation is dangerous and may operate an unexpected k8s cluster, so we should avoid automatically finding the kubeconfig file in the user's home directory to create the k8s client.

Reply:
This is same with someone execute `kubectl delete sth.` without confirming the context first, IMHO

Reply:
It makes sense to avoid operating on the unexpected k8s cluster, but even if users need to explicitly specify `--kubeconfig ~/.kube/config`, they may run against the wrong cluster because they forgot to switch context. Maybe we can ask for confirmation or/both add some preflight checks?

- show cluster information and ask the user for confirmation before proceeding (and add `--skip-confirm` to skip the confirmation explicitly in the batch job)
- add some preflight checks, e.g. check the version of tidb operator matches expected version

Reply:
This is completely different from someone executes a dangerous command like `kubectl delete sth`, This is the user's own behavior, and your program may perform dangerous operations without the user's perception.

Reply:
@cofyc @onlymellb I agree the safety concern, however, there is always a trade-off between convenient, safety and complexity. I think a reasonable solution for now is remind user of this in the documetation, how do you think?

Reply:
After all, out-of-cluster stability test is for convenient "stability test developing" but not actually running.

Reply:
I just don't think it's necessary to automatically find the kubeconfig file in the user's home directory to create the k8s client.  If the user does not specify kubeconfig explicitly and does not run the program in the k8s cluster, let it exit directly. What do you think？ @weekface @tennix @xiaojingchen

Reply:
I agree with @onlymellb

Reply:
emmm... I obey the majority then